Add ModificationType enum and version id, rename MsgDeleted

daehn opened the pull request:
This PR adds a new `ModificationType` enum to the spec. If a message needs to be edited or deleted (on all devices), the optional `modification` field needs to be set and the message. 
We introduced a new field `version_id` which should be set to a time based UUID when inserting a message as well as when applying a modification. Updates to messages (like adding a link preview for example) are only valid if they share the same `version_id` as the original message. Using the `version_id` we can avoid race conditions, for example when a message containing a link gets edited before the link preview was fetched and sent, which then might no longer be the correct preview. No `version_id` should always be considered the oldest version to keep compatibility with older clients.

zbsz commented:
1. Do we actually need `version_id` in proto? Every message has a server timestamp which should be enough to resolve any conflicts. Mentioned race condition seems like client specific thing that can be resolved locally.
2. `modification` is only relevant for specific message types, ie. only text messages can be edited, so having that as top level property seems too much. Instead of that we could add specific message content, something like: `MsgRecall { msg_id }` and `MsgEdit { msg_id, updated_content: Text }`

bennyn commented:
Do be consistent we shouldn't shorten "Message" to "Msg" because we have already a `message GenericMessage`. 

So I propose to call it `message MessageHide` instead of `message MsgHide`.

daehn commented:
I updated the PR regarding the comments, the delete / recall action has been extracted to be it's own message type.

I removed the enum as a modification can be implied by a changed `version_id`. The `version_id` should only be changed if the original content of the message has been changed (edit), not if a message gets updated with a `LinkPreview` or `Asset` message.

daehn commented:
As they are some problems with using a type 1 UUID (e.g. not in sync device clocks) which is why we split up the versioning logic in two separate fields `version_id` and `version_number`. 

If a message gets modified (edited) the `version_number` should be increased by one and a new random `version_id` generated. If two clients do this at the same time their version will be distinguishable using the `version_id`, even if they have the same `version_number`, it can then be decided randomly which one wins.

daehn commented on proto/messages.proto:
```diff
@@ -14,9 +14,13 @@ message GenericMessage {
   }
+
+  optional string version_id = 15;
+  optional int32 version_number = 16;
```
It might make more sense to extract this in a separate message to ensure that both fields are always set if the message has a version

zbsz replied:
good idea

zbsz commented:
Could you also add documentation in README. Please specify how the conflict resolution is done as all clients should use the same logic to make sure the same message 'wins' on every device.

zbsz commented:
👍

zbsz commented:
One more issue with version based edit: it will conflict with local delete. If client deletes a message locally, and then receives an edit update, then it shouldn't show this message. We would need to rely on version numbers to check if that is en edit or not. It would be nicer to have this information explicitly.

daehn commented:
That's right, good catch. I would then suggest adding an additional `edited` flag to the `GenericMessage` type, it can default to false and we don't limit the spec to only text messages being editable?

daehn commented:
As discussed there have been some changes: I removed the `Version` and added a new type `MessageEdit` which references another nonce and contains the edited content. For now only content of type `Text` is supported. If a message of type `MessageEdit` is received, the nonce and the content of the referenced message should be updated . If the referenced message is not present (because it has been edited in the meantime or has been deleted) the edit should be discarded.

zbsz commented:
@daehn could you please squash commits, no need to keep all that history :)

daehn commented:
@zbsz done
